File: rb/happo.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

BEGIN {
  print "はじまるよ\n"
}

END {
  print "おわるよ\n"
}

# --------------------------------------------------------------------
# 入出力。Ruby のブロックは、リソースのスコープが一目瞭然ですばらしいな

def io()
  filename = "/tmp/ruby-test"
  open(filename, "w") { |output|
    output.write("Hello, World!\n")
    output.write("Foo Bar\n")
  }
  IO.popen("/bin/cat -n #{filename}") { |input|
    #while (line = input.gets)
    #  print line
    #end
    input.readlines.each { |line|
      print line
    }
  }
end

# --------------------------------------------------------------------
# 制御構造は式になる。primitive 以外は method + block を使った方が綺麗

def control_structure
  if true then
    ;
  elsif false then
    ;
  else
    ;
  end
  loop { # これは新スコープ。loop は、ブロック付きで呼び出されるメソッド
    print "Looping in a block.\n"
    break
  }
  while true do # while true do はスコープを作らない。いわゆる制御構造
    print "Looping with control structure.\n"
    break
  end
  (1 .. 3).each { |i|
    print i.to_s + "\n"
  }
  for i in ["A", "B", "C",] do
    print i << "\n"
  end
  # 後置できます
  print "postpisition\n" if true
  i = 0
  for i in (0 .. 9) do
    # switch できます。
    case i + 1
    when 0 .. 2 then
      i += 1
      p "foo"
      redo # 条件チェックなしでもう一度
    when 3 .. 4, 5 .. 6 then
      p "bar"
      next
    else
      ;
    end
    p i
  end
  # Ruby の「イテレータ」とは、ぶっちゃけ yield を使うメソッドのこと
  def my_while(cond, s)
    return if ! cond
    yield s # クロージャ呼び出し
    retry # イテレータ引数を再評価し、イテレータを再実行
  end
  i = 0
  # なるほど、たしかにRubyのブロックはクロージャだ。イテレータ
  # my_while() はメソッドだから呼び出し元のスタックにはアクセスできない
  # けれど、そこからyieldで呼ばれるブロックはクロージャだからアクセスし
  # 放題なわけだ。
  my_while(i < 10, "hoge") { |s|
    i = i + 1
    if i == 5 then
      next # yield 脱出
    end
    # Python 式の format も使えるのか
    print "%s: %d\n" % [s, i,]
  }
  # 3 項演算はこんな感じ。制御構造も式
  print "result0: " << ((true)? "TRUE": "FALSE") << "\n"
  print "result1: " << (if true then "TRUE" else "FALSE" end) << "\n"
  # 例外処理
  begin
    raise "hoge"
  rescue Exception then
    p "cp0"
  rescue
    p "cp1"
  else
    p "cp2"
  ensure # いわゆる finally か
    p "cp3"
  end
end

control_structure

#if true then; a = 123; end; print a # OK 
#loop { b = 123; break; }; print b # NG

# --------------------------------------------------------------------
# データ操作
def data_manipulation
  # 文字列操作
  # http://www.ruby-lang.org/ja/man/html/String.html
  s = "01234" + "56789"
  print "#{s[2 .. 8]}, #{s.length}\n"
  print "#{s[2 .. -2]}, #{s.length}\n"
  print "#{s[2, 7]}, #{s.length}\n"
  print "d: #{s}\n"
  print "d: #{s[/0+([1-2]+)3+([4-5]+)6+([7-8]*)9/, 2]}\n"
  # リスト操作
  # http://www.ruby-lang.org/ja/man/html/Array.html
  a = [0, 1, 2, 3, 4, 5,]
  a << 6 # a.push(6) でも同じ
  a.pop()
  a += [7, 8, 9,]
  p a[2 .. 8], a.size # length でも可
  # 連想配列のような、ペアの検索もできる
  d = [[2, 234,], [3, 345,], [4, 456,],]
  p d.assoc(3)
  # 連想配列 (ハッシュ)
  #d = Hash[
  # *[
  #  2, 2 ** 2,
  #  3, 3 ** 2,
  # ]
  #]
  d = {
   2 => 2 ** 2,
   3 => 3 ** 2,
   :id0 => "hoge",
   :id1 => "fuga",
   "id0" => "HOGE",
   "id1" => "FUGA",
  }
  d[9] = "d9"
  # each_key や each_value では、それぞれ
  d.each { |key, val|
    print "d: #{key.inspect} #{val}\n"
  }
  if d.has_key?(2) then
    print "#{d.fetch(2, "not_found")}\n"
  end
  # 関数ポインタの扱い
  def foo(s)
    print "Called #{s}\n"
  end
  m = :main.method("foo") # :foo でも可だが、
   # 多分中では結局文字列比較してるだろう
  m.call("Hello")
end

data_manipulation

# --------------------------------------------------------------------
# 正規表現

def re_operation()
  # Regexp 型リテラル
  re_test = /\((?:b\w)(dy)\)\s/
  print "re_operation: "
  p re_test
  re_test =~ "head-(body) tail"
  p $~
  if (m = re_test.match("head-(body) tail")) then
    p m.to_a
  end
end

re_operation

# --------------------------------------------------------------------
# オブジェクト指向プログラミング



# --------------------------------------------------------------------
# ?

# --------------------------------------------------------------------
# 関数とスコープ

# ローカル変数
# インスタンス変数
# クラス変数
# グローバル変数
# 疑似変数
# 定数

ConstantValue = 123 # 大文字で始まる

# Python で言うところの dir()
s = "string"
print "%s\n" % [s.methods.sort.join(" "),]

File: py/happo.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# encoding declatation <http://www.python.jp/doc/release/ref/encodings.html>

# --------------------------------------------------------------------
# 入出力

def io():
  import tempfile

  # このテンポラリは、閉じると消えてしまう。テンポラリは mkdtemp() 内が吉
  #output = tempfile.NamedTemporaryFile("w", suffix = ".suffix")
  #filename = output.name

  #filename = tempfile.mktemp(".postfix")
  ## 普通に開く
  #output = open(filename, "w")

  fd, filename = tempfile.mkstemp(".postfix")
  import os
  output = os.fdopen(fd, "w")

  output.write("Hello, World!\n")
  output.write("Foo Bar!\n")
  output.close()
  import sys
  sys.stdout.write("Filename: %s\n" % (filename,))
  # popen する
  import os
  input = os.popen("/bin/cat -n %s" % (filename,), "r")
  # 一行ずつ読み出す、標準的な方法。逐次なら readline()
  for line in input.readlines():
    #line = line.rstrip()
    print line,
  input.close()

io()

# --------------------------------------------------------------------
# 制御構造

def control_structure():
  if True:
    pass
  elif False:
    pass
  else:
    pass
  while True:
    print "Looping"
    break
  for i in range(3):
    print i
  # ----
  # No DO WHILE
  i = 0
  n = 5
  while True:
    pass
    if i > n:
      break
    i += 1;
  # ----
  # No post-positioned IF
  # 前置でいいよ
  if True: print "Hoge"
  # ----
  # No SWITCH
  # switch は、コーディング時に条件と挙動が固定化されてしまうので、動的
  # なプログラミングにならず、仕様変更に弱いのだ
  def print_abc():
    print "ABC"
  key = "abc"
  tbl = {"abc": print_abc}
  tbl.get(key)()
  def print_xyz():
    print "XYZ"
  import sys
  tbl["xyz"] = lambda: sys.stdout.write("XYZ\n")
  tbl.get("xyz")()
  # ----
  # 3 項演算子もないが…
  print ("FALSE", "TRUE",)[True]
  # ----
  # 例外処理 (Exception)
  try:
    a = not_exist[0]
  except NameError, e:
    # 例外時
    print "***", e
  except:
    pass
  else:
    pass
  finally:
    # 後処理
    pass

control_structure()

# --------------------------------------------------------------------
# データ操作

def data_manipulation():
  # 文字列操作
  s = "01234" + "56789"
  print s[2:8], len(s)
  # リストの操作
  a = [0, 1, 2, 3, 4, 5,]
  a.append(6)
  a.pop()
  a.append(7)
  a.append(8)
  a.append(9)
  print a[2:8], len(a)
  # todo: 内包表現も
  # 連想配列
  d = {0: "d0", 1: "d1", 2: "d2",}
  d[9] = "d9"
  # .keys() は、無くとも良い
  if 2 in d.keys():
    print d.get(8, "not_found"), len(d)
  # 関数型の扱い
  def func():
    print "dynamic funciton call"
  # Python では、変数と関数は名前空間を共有する
  p = func
  p()

data_manipulation()

# --------------------------------------------------------------------
# 正規表現

def re_operation():
  # http://www.python.jp/doc/nightly/lib/module-re.html
  import re
  # \w = [a-zA-Z0-9_]
  # (?:～) は、参照できないグループ
  # (?P<key>val) は、名前による参照グループ
  re_test = re.compile("\((?:b\w)(?P<foo>dy)\)\s")
  m = re_test.search("head-(body) tail")
  if m:
    print m.group(0)
    print m.groups()
    print m.groupdict()

re_operation()

# --------------------------------------------------------------------
# オブジェクト指向プログラミング

class Hoge(object):
  def __init__(self):
    print "Hoge init"
  def foo(self):
    print "foo"

class Fuga(Hoge):
  def __init__(self):
    print "Fuga init"
    super(type(self), self).__init__()
  def foo(self):
    print "bar"
    super(Fuga, self).foo()

def oop():
  fuga = Fuga()
  fuga.foo()

oop()

# --------------------------------------------------------------------


#  モジュール化, require とか
#  class
#  継承
#  private/public

# --------------------------------------------------------------------
# 関数とスコープ

global_value0 = "global value0 unchanged"
global_value1 = "global value1 unchanged"

# def scope0():
#   print "local0:", global_value0 # 参照は OK だが、これはグローバルなのに、
#   global_value0 = "global value0 changed" # これがローカルなのでエラー。

def scope0():
  global_value0 = "global value0 changed" # これはローカルで、
  print "func:", global_value0 # これもローカルだから OK。

def scope1():
  global global_value1
  global_value1 = "global value1 changed"
  print "func:", global_value1

print "pre:", global_value0
scope0()
print "after:", global_value0

print "pre:", global_value1
scope1()
print "after:", global_value1

# --------------------------------------------------------------------

def python_specific():
  pass


python_specific()
